linalg/kernels/densemat.cpp

```diff
@@ -0,0 +1,359 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../../general/okina.hpp"
+
+namespace mfem
+{
+
+// *****************************************************************************
+template <class T> __device__ __host__
+inline void Swap(T &a, T &b)
+{
+   T tmp(a);
+   a = b;
+   b = tmp;
+}
+
+// *****************************************************************************
+void kLSolve( const int m,
+              const int n,
+              const double *data, const int *ipiv, double *x)
+{
+   GET_CONST_ADRS(data);
+   GET_CONST_ADRS_T(ipiv,int);
+   GET_ADRS(x);
+   MFEM_FORALL(k, n,
+   {
+      double *d_mx = &d_x[k*m];
+      // X <- P X
+      for (int i = 0; i < m; i++)
+      {
+         Swap<double>(d_mx[i], d_mx[d_ipiv[i]]);
+      }
+      // X <- L^{-1} X
+      for (int j = 0; j < m; j++)
+      {
+         const double d_mx_j = d_mx[j];
+         for (int i = j+1; i < m; i++)
+         {
+            d_mx[i] -= d_data[i+j*m] * d_mx_j;
+         }
+      }
+   });
+}
+
+// *****************************************************************************
+void kUSolve(const int m, const int n, const double *data, double *x)
+{
+   GET_CONST_ADRS(data);
+   GET_ADRS(x);
+   MFEM_FORALL(k, n,
+   {
+      double *d_mx = &d_x[k*m];
+      for (int j = m-1; j >= 0; j--)
+      {
+         const double x_j = ( d_mx[j] /= d_data[j+j*m] );
+         for (int i = 0; i < j; i++)
+         {
+            d_mx[i] -= d_data[i+j*m] * x_j;
+         }
+      }
+   });
+}
+
+// *****************************************************************************
+void kFactorPrint(const int s, const double *data)
+{
+   GET_CONST_ADRS(data);
+   MFEM_FORALL(i, s,
+   {
+      printf("\n\td_data[%ld]=%f",i,d_data[i]);
+   });
+}
+
+// *****************************************************************************
+void kFactorSet(const int s, const double *adata, double *ludata)
+{
+   GET_CONST_ADRS(adata);
+   GET_ADRS(ludata);
+   MFEM_FORALL(i, s,
+   {
+      d_ludata[i] = d_adata[i];
+   });
+}
+
+// *****************************************************************************
+void kFactor(const int m, int *ipiv, double *data)
+{
```

Comment:
@camierjs , I don't think this LU factorization routine is valid on the GPU as is. I believe this is inherently a sequential algorithm. I get different answers when looking at the CPU vs GPU.